worker/sshserver/server.go

```
@@ -66,48 +77,40 @@ func NewServerWorker(config ServerWorkerConfig) (worker.Worker, error) {
 		},
 	}
 
-	// TODO(ale8k): Update later to use the host key from StateServingInfo
-	terminatingHostKey, err := rsa.GenerateKey(rand.Reader, 2048)
-	if err != nil {
-		return nil, errors.Annotate(err, "failed to generate host key")
-	}
-	signer, err := gossh.NewSignerFromKey(terminatingHostKey)
-	if err != nil {
-		return nil, errors.Annotate(err, "failed to create signer")
+	// Set hostkey.
+	if err := s.setJumpServerHostKey(); err != nil {
+		return nil, errors.Trace(err)
 	}
 
-	// Set hostkey
-	s.Server.AddHostKey(signer)
-
 	if s.config.Listener == nil {
-		listenAddr := net.JoinHostPort("", strconv.Itoa(2229))
+		listenAddr := net.JoinHostPort("", strconv.Itoa(config.Port))
 		listener, err := net.Listen("tcp", listenAddr)
 		if err != nil {
 			return nil, err
 		}
 		s.config.Listener = listener
 	}
 
-	listener := newAcceptOnceListener(s.config.Listener)
+	listener, closeAllowed := NewSSHServerListener(s.config.Listener)
 
-	// Start server
+	// Start server.
 	s.tomb.Go(func() error {
 		err := s.Server.Serve(listener)
 		if errors.Is(err, ssh.ErrServerClosed) {
 			return nil
 		}
-		return err
+		return errors.Trace(err)
```

This error is being ignored. You can see this by hacking it to return an error and tests still pass etc.

Plus, calling tomb.Go again below is bad. 
```
// Calling the Go method after all tracked goroutines return
// causes a runtime panic. For that reason, calling the Go
// method a second time out of a tracked goroutine is unsafe.
```

I'm not convinced we need 2 workers. Just one will be enough to watch for config changes and kill and restart the ssh server when the config changes.

---

It isn't ignored:
```
FAIL: server_test.go:82: sshServerSuite.TestSSHServer

server_test.go:143:
    // Server isn't gracefully closed, it's forcefully closed. All connections ended
    // from server side.
    workertest.CleanKill(c, server)
/home/joseph/go/pkg/mod/github.com/juju/worker/v3@v3.5.0/workertest/check.go:83:
    c.Check(err, jc.ErrorIsNil)
... value *errors.Err = &errors.unformatter{message:"", cause:(*errors.Err)(0xc000244000), previous:(*errors.Err)(0xc000244000), function:"github.com/juju/juju/worker/sshserver.NewServerWorker.func3", line:100} ("FAAARK")
... error stack:
        github.com/juju/juju/worker/sshserver.NewServerWorker.func3:100: FAAARK
        github.com/juju/juju/worker/sshserver.NewServerWorker.func3:100:

ignoring error: FAAARK
```

Also, those `tomb.Go` calls are inside the main worker's loop, which is invoked by `tomb.Go`. This means that to get to a zero count inside the tomb, the main work has to exit, at which point it is semantically impossible to call `tomb.Go` again.

---

I tested it and didn't get that output. I was surprised. But clearly was holding it wrong.